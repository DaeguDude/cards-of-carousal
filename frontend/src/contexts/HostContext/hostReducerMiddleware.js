import socketInstance from '../../socket/socket';

function closeGame() {
  socketInstance.closeSocket();
}

function createLobby() {
  socketInstance.createLobby();
}

function sendPlayerConnectedMessage(payload) {
  socketInstance.sendMessage({
    event: 'update',
    recipients: [payload.playerId],
    payload: {
      gameState: 'connected',
      message: {
        big: "You've joined the lobby",
        small: 'Please wait for the host to start the game',
      },
    },
  });
}

function sendKickPlayerMessage(payload) {
  socketInstance.sendMessage({
    recipients: [payload.playerId],
    event: 'update',
    payload: {
      message: {
        big: "You've been kicked!",
        small: 'Take off, you hoser!',
      },
    },
  });
}

async function getDeck({ selectedPacks }) {
  try {
    const apiURL = process.env.REACT_APP_API_URL;
    const query = selectedPacks.join(',');
    const cardsRequest = await fetch(`${apiURL}/deck/cards?packs=${query}`);
    const cards = await cardsRequest.json();
    return cards;
  } catch (err) {
    throw new Error(err);
  }
}

export default async function hostReducerMiddleware(
  { type, payload },
  dispatch,
) {
  switch (type) {
    case `CLOSE_GAME`:
      closeGame();
      break;

    case 'CREATE_LOBBY':
      createLobby();
      break;

    case 'PLAYER_CONNECTED':
      sendPlayerConnectedMessage(payload);
      break;

    case 'KICK_PLAYER':
      sendKickPlayerMessage(payload);
      break;

    case 'SET_DECK': {
      const deck = await getDeck(payload);
      return dispatch({
        type: 'SET_DECK',
        payload: { deck },
      });
    }

    default:
      break;
  }
  return dispatch({ type, payload });
}
