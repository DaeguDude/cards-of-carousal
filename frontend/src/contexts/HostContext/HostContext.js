import React, { createContext, useEffect } from 'react';
import PropTypes from 'prop-types';

import useReducerMiddleware from '../useReducerMiddleware';
import HostReducer from './HostReducer';
import socketInstance from '../../socket/socket';
import hostReducerMiddleware from './hostReducerMiddleware';

const propTypes = {
  children: PropTypes.node.isRequired,
};

const { emitter } = socketInstance;

const initialState = {
  gameState: 'waiting-for-lobby',
  lobbyID: '',
  players: {},
  playerIDs: [],
  gameSettings: { maxPlayers: 8, winningScore: 7, selectedPacks: [] },
};

export const HostContext = createContext();

function HostProvider({ children }) {
  const [state, dispatch] = useReducerMiddleware(
    hostReducerMiddleware,
    HostReducer,
    initialState,
  );

  function handleMessage({ event, payload, sender }) {
    switch (event) {
      case 'player-connected':
        return dispatch({ type: 'PLAYER_CONNECTED', payload });
      case 'player-disconnected':
        return dispatch({ type: 'PLAYER_DISCONNECTED', payload });
      case 'select-winner':
        return dispatch({ type: 'SELECT_WINNER', payload });
      case 'lobby-created':
        return dispatch({ type: 'SET_LOBBY_ID', payload });
      case 'player-submit':
        return dispatch({
          type: 'PLAYER_SUBMIT',
          payload: { ...payload, playerId: sender },
        });
      default:
        return undefined;
    }
  }

  useEffect(() => {
    emitter.on('message', handleMessage);
    return () => {
      emitter.off('message', handleMessage);
    };
  }, []);

  return (
    <HostContext.Provider value={{ state, dispatch }}>
      {children}
    </HostContext.Provider>
  );
}

HostProvider.propTypes = propTypes;

export default HostProvider;
