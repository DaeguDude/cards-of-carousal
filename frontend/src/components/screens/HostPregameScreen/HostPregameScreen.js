import React, { useContext, useEffect } from 'react';
import styled from 'styled-components';
import { HostContext } from '../../../contexts/HostContext/HostContext';
import HostLayout from '../../layouts/HostLayout';
import PlayerList from '../../PlayerList/PlayerList';
import GameSettings from '../../GameSettings/GameSettings';
import HostSettingsMenu from '../../HostSettingsMenu/HostSettingsMenu.js';
import JoinCode from '../../JoinCode/JoinCode';
import Button from '../../Buttons/Button';

const LeftPanelWrapper = styled.div`
  display: flex;
  flex-direction: column;
  height: 100%;

  .player-list-wrapper {
    overflow-x: auto;
    border-bottom: 1px solid var(--primary-text-color);
    flex-grow: 1;
  }

  .buttons-wrapper {
    display: block;
    margin-bottom: 20px;
  }

  .bottom-left-wrapper {
    margin: 20px auto 32px;
    position: relative;
    bottom: 0;
    left: 0;
    right: 0;
  }

  .join-code-wrapper {
    display: flex;
    justify-content: center;
  }

  .host-pregame-button {
    padding: 0;
    background-color: var(--primary-background-color);
    color: var(--primary-text-color);
    margin: auto auto 12px;
  }

  .start-carousing {
    font-size: 0.875rem;
    padding: 16px 34px;
  }

  .close-game {
    font-size: 0.625rem;
    padding: 8px 20px;
  }
`;

const RightPanelWrapper = styled.div`
  display: flex;
  flex-direction: column;
  position: absolute;
  top: 0;
  bottom: 0;
  right: 0;
  left: 0;

  .game-description {
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    background-color: var(--primary-background-color);
    align-self: center;
    margin: 24px;
    padding: 8px;
    border-radius: 5px;
    font-size: 1.8rem;
    font-weight: 700;
    box-shadow: 0px 4px 12px rgba(0, 0, 0, 0.25);
  }

  .game-description p {
    margin: 16px 0;
  }

  .top {
    display: flex;
    height: 60%;
  }

  .bottom {
    height: 40%;
  }
`;

function LeftPanel() {
  const { state, dispatch } = useContext(HostContext);
  const { players, playerIDs, lobbyID } = state;

  const handleClickStart = () => {
    // check if there are any players
    if (Object.keys(state.players).length) {
      dispatch({
        type: 'SET_GAME_STATE',
        payload: { gameState: 'waiting-for-deck' },
      });
      dispatch({
        type: 'SET_NEW_CZAR',
        payload: {},
      });
    }
    // TODO: add else statement to warn that you cannot play a game with no players
  };

  const handleClickClose = () => {
    dispatch({ type: 'CLOSE_GAME' });
    window.location.reload();
  };

  return (
    <LeftPanelWrapper>
      <div className="player-list-wrapper">
        <PlayerList playerList={{ players, playerIDs }} />
      </div>
      <div className="bottom-left-wrapper">
        <div className="buttons-wrapper">
          <Button className="host-pregame-button" onClick={handleClickStart}>
            <div className="start-carousing">START CAROUSING</div>
          </Button>
          <Button className="host-pregame-button" onClick={handleClickClose}>
            <div className="close-game">CLOSE GAME</div>
          </Button>
        </div>
        <div className="join-code-wrapper">
          <JoinCode code={lobbyID} />
        </div>
      </div>
    </LeftPanelWrapper>
  );
}

function RightPanel() {
  const { state, dispatch } = useContext(HostContext);
  const { gameSettings } = state;

  const onChangeSettings = (settings) => {
    dispatch({
      type: 'SET_GAME_SETTINGS',
      payload: { gameSettings: settings },
    });
  };

  return (
    <RightPanelWrapper>
      <div className="top">
        <div className="game-description">
          <p>Cards of Carousal is a game for lorem ipsum dolor.</p>
          <p>
            Lorem ipsum dolor sit amet consectetur adipisicing elit. Esse,
            doloribus quaerat incidunt excepturi odit eos!
          </p>
          <p>
            Qui delectus laboriosam aperiam maxime optio, architecto asperiores,
            at ullam.
          </p>
        </div>
      </div>
      <div className="bottom">
        <GameSettings options={gameSettings} onChange={onChangeSettings} />
      </div>
    </RightPanelWrapper>
  );
}

function HostPregameScreen() {
  const { dispatch } = useContext(HostContext);

  useEffect(() => {
    dispatch({ type: 'CREATE_LOBBY', payload: {} });
  }, []);

  return (
    <HostLayout
      className="primary-background"
      left={<LeftPanel />}
      right={<RightPanel />}
      modal={<HostSettingsMenu />}
    />
  );
}

export default HostPregameScreen;
