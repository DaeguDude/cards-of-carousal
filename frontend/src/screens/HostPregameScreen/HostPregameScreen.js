import React, { useContext, useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { HostContext } from '../../contexts/HostContext/HostContext';
import HostLayout from '../../layouts/HostLayout';
import PlayerList from '../../components/PlayerList/PlayerList';
import GameSettings from '../../components/GameSettings/GameSettings';
import PregameSettingsModal from '../../components/HostSettingsMenu/PregameSettingsModal';
import JoinCode from '../../components/JoinCode/JoinCode';
import Button from '../../components/Buttons/Button';
import LoadingIndicator from '../../components/LoadingIndicator/LoadingIndicator';
import AlertModal from '../../components/Modal/AlertModal';

const errorHandlerPropTypes = {
  errorString: PropTypes.string.isRequired,
  setError: PropTypes.func.isRequired,
};

const LeftPanelWrapper = styled.div`
  display: flex;
  flex-direction: column;
  height: 100%;

  .player-list-wrapper {
    overflow-x: auto;
    border-bottom: 1px solid var(--primary-text-color);
    flex-grow: 1;
  }

  .buttons-wrapper {
    display: block;
    margin-bottom: 20px;
  }

  .bottom-left-wrapper {
    margin: 20px auto 32px;
    position: relative;
    bottom: 0;
    left: 0;
    right: 0;
  }

  .join-code-wrapper {
    display: flex;
    justify-content: center;
  }

  .host-pregame-button {
    padding: 0;
    background-color: var(--primary-background-color);
    color: var(--primary-text-color);
    margin: auto auto 12px;
  }

  .start-carousing {
    font-size: 0.875rem;
    padding: 16px 34px;
  }

  .close-game {
    font-size: 0.625rem;
    padding: 8px 20px;
  }
`;

const RightPanelWrapper = styled.div`
  display: flex;
  flex-direction: column;
  position: absolute;
  top: 0;
  bottom: 0;
  right: 0;
  left: 0;

  .game-description {
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    background-color: var(--primary-background-color);
    align-self: center;
    margin: 24px;
    padding: 8px;
    border-radius: 5px;
    font-size: 1.8rem;
    font-weight: 700;
    box-shadow: 0px 4px 12px rgba(0, 0, 0, 0.25);
  }

  .game-description p {
    margin: 16px 0;
  }

  .top {
    display: flex;
    justify-content: center;
    align-items: center;
    height: 60%;
  }

  .bottom {
    height: 40%;
  }
`;

function ErrorHandler({ errorString, setError }) {
  function reload() {
    return window.location.reload();
  }
  function closeModal() {
    return setError('');
  }

  switch (errorString) {
    case 'error-getting-cards':
      return (
        <AlertModal
          bigText="Failed to retrieve cards"
          smallText="Please try again later"
          buttonText="Click to restart"
          onClick={reload}
        />
      );
    case 'not-enough-players':
      return (
        <AlertModal
          bigText="Unable to start game"
          smallText="No offense, but this game requires friends to play."
          buttonText="Click anywhere to continue"
          onClick={closeModal}
        />
      );
    case 'no-card-packs-selected':
      return (
        <AlertModal
          bigText="Unable to start game"
          smallText="Please pick at least one card pack."
          buttonText="Click anywhere to continue"
          onClick={closeModal}
        />
      );
    default:
      return null;
  }
}

ErrorHandler.propTypes = errorHandlerPropTypes;

function LeftPanel() {
  const { state, dispatch } = useContext(HostContext);
  const { players, playerIDs, lobbyID } = state;
  const [error, setError] = useState('');

  const handleClickStart = async () => {
    // check if there are any players and if packs are selected
    if (playerIDs.length > 1 && state.gameSettings.selectedPacks.length) {
      const { selectedPacks } = state.gameSettings;

      dispatch({ type: 'GET_DECK', payload: {} });

      try {
        await dispatch({
          type: 'SET_DECK',
          payload: { selectedPacks },
        });

        dispatch({
          type: 'START_GAME',
          payload: {},
        });

        dispatch({
          type: 'SET_NEXT_CZAR',
          payload: {},
        });

        dispatch({
          type: 'SELECT_BLACK_CARD',
          payload: {},
        });

        dispatch({ type: 'DEAL_WHITE_CARDS', payload: {} });
      } catch (err) {
        // The only cause of an error here would be a failure to retrieve card packs
        setError('error-getting-cards');
      }
    } else {
      // The only failure cases here are:
      //  a) not enough players, or b) card packs aren't selected
      const errorString =
        playerIDs.length > 1 ? 'no-card-packs-selected' : 'not-enough-players';

      setError(errorString);
    }
  };

  const handleClickClose = async () => {
    await dispatch({ type: 'CLOSE_GAME', payload: {} });
    window.location.reload();
  };

  return (
    <LeftPanelWrapper>
      <div className="player-list-wrapper">
        <PlayerList playerList={{ players, playerIDs }} />
      </div>
      <div className="bottom-left-wrapper">
        <div className="buttons-wrapper">
          <Button className="host-pregame-button" onClick={handleClickStart}>
            <div className="start-carousing">START CAROUSING</div>
          </Button>
          <Button className="host-pregame-button" onClick={handleClickClose}>
            <div className="close-game">CLOSE GAME</div>
          </Button>
        </div>
        <div className="join-code-wrapper">
          <JoinCode
            loading={state.loading.includes('join-code')}
            code={lobbyID}
          />
        </div>
      </div>

      <ErrorHandler errorString={error} setError={setError} />
    </LeftPanelWrapper>
  );
}

function RightPanel() {
  const { state, dispatch } = useContext(HostContext);
  const { gameSettings } = state;

  const onChangeSettings = (settings) => {
    dispatch({
      type: 'SET_GAME_SETTINGS',
      payload: { gameSettings: settings },
    });
  };

  return (
    <RightPanelWrapper>
      <div className="top">
        {state.loading.includes('getting-deck') ? (
          <LoadingIndicator />
        ) : (
          <div className="game-description">
            <p>Cards of Carousal is a game for lorem ipsum dolor.</p>
            <p>
              Lorem ipsum dolor sit amet consectetur adipisicing elit. Esse,
              doloribus quaerat incidunt excepturi odit eos!
            </p>
            <p>
              Qui delectus laboriosam aperiam maxime optio, architecto
              asperiores, at ullam.
            </p>
          </div>
        )}
      </div>
      <div className="bottom">
        <GameSettings options={gameSettings} onChange={onChangeSettings} />
      </div>
    </RightPanelWrapper>
  );
}

function HostPregameScreen() {
  const { dispatch } = useContext(HostContext);

  useEffect(() => {
    dispatch({ type: 'CREATE_LOBBY', payload: {} });
  }, []);

  return (
    <HostLayout
      className="primary-background"
      left={<LeftPanel />}
      right={<RightPanel />}
      modal={<PregameSettingsModal />}
    />
  );
}

export default HostPregameScreen;
