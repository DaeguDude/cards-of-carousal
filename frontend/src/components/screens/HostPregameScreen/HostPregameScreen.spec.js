import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { HostContext } from '../../../contexts/HostContext/HostContext';
import HostPregameScreen from './HostPregameScreen';

describe('Host Pregame Screen', () => {
  // This is the default state value provided by our context
  let state = {
    gameState: 'waiting-for-lobby',
    lobbyID: '',
    players: {},
    playerIDs: [],
    gameSettings: { maxPlayers: 8, winningScore: 7, selectedPacks: [] },
  };
  const dispatch = jest.fn();

  afterEach(() => {
    // sometimes we need to change the state for test cases, this resets to initial value
    state = {
      gameState: 'waiting-for-lobby',
      lobbyID: '',
      players: {},
      playerIDs: [],
      gameSettings: { maxPlayers: 8, winningScore: 7, selectedPacks: [] },
    };
  });

  describe('rendering', () => {
    it('renders', () => {
      const tree = renderer
        .create(
          <HostContext.Provider value={{ state, dispatch }}>
            <HostPregameScreen />
          </HostContext.Provider>,
        )
        .toJSON();

      expect(tree).toMatchSnapshot();
    });
  });

  describe('lobby', () => {
    it('creates the lobby', () => {
      expect(dispatch).not.toHaveBeenCalled();

      render(
        <HostContext.Provider value={{ state, dispatch }}>
          <HostPregameScreen />
        </HostContext.Provider>,
      );

      expect(dispatch).toHaveBeenCalledTimes(1);
      expect(dispatch.mock.calls[0][0]).toEqual({
        type: 'CREATE_LOBBY',
        payload: {},
      });
    });
  });

  describe('buttons', () => {
    it('reloads when the close game button is clicked', () => {
      // window.location properties are read-only, we have to redefine this object to spy on reload
      Object.defineProperty(window, 'location', {
        value: { reload: jest.fn() },
      });
      render(
        <HostContext.Provider value={{ state, dispatch }}>
          <HostPregameScreen />
        </HostContext.Provider>,
      );

      fireEvent.click(screen.getByText('CLOSE GAME'));

      expect(window.location.reload).toHaveBeenCalledTimes(1);
      // create lobby, close game
      expect(dispatch).toHaveBeenCalledTimes(2);
      expect(dispatch.mock.calls[1][0]).toEqual({
        type: 'CLOSE_GAME',
        payload: {},
      });
    });

    it('calls a dispatch when the start button is pressed', () => {
      state = {
        gameState: 'waiting-for-lobby',
        lobbyID: '',
        players: {
          foo: {
            name: 'Bender',
            score: 0,
            isCzar: false,
            submittedCards: [],
            cards: [],
          },
          bar: {
            name: 'Briggs',
            score: 0,
            isCzar: false,
            submittedCards: [],
            cards: [],
          },
          baz: {
            name: 'Pedro',
            score: 0,
            isCzar: false,
            submittedCards: [],
            cards: [],
          },
        },
        playerIDs: [],
        gameSettings: { maxPlayers: 8, winningScore: 7, selectedPacks: [] },
      };

      render(
        <HostContext.Provider value={{ state, dispatch }}>
          <HostPregameScreen />
        </HostContext.Provider>,
      );

      fireEvent.click(screen.getByText('START CAROUSING'));

      // FIX ME: create lobby, set game state, set new czar, get deck
      expect(dispatch).toHaveBeenCalledTimes(4);
      expect(dispatch.mock.calls[1][0]).toEqual({
        type: 'START_GAME',
        payload: {},
      });
      expect(dispatch.mock.calls[2][0]).toEqual({
        type: 'SET_NEXT_CZAR',
        payload: {},
      });
    });

    it('does not call dispatch if no players are in the lobby when the start button is clicked', () => {
      render(
        <HostContext.Provider value={{ state, dispatch }}>
          <HostPregameScreen />
        </HostContext.Provider>,
      );

      fireEvent.click(screen.getByText('START CAROUSING'));

      // only create lobby called
      expect(dispatch).toHaveBeenCalledTimes(1);
    });
  });
});
