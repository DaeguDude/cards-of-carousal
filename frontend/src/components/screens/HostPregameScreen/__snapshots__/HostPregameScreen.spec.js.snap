// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Host Pregame Screen rendering renders 1`] = `
.c1 {
  background-image: url(inactiveHamburger.svg);
  background-size: cover;
  background-color: transparent;
  border: 0;
  width: 25px;
  height: 25px;
}

.c2 {
  width: 100%;
  background-color: var(--secondary-background-color);
  color: var(--secondary-text-color);
}

.c0 {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 .hamburger-container {
  position: absolute;
  right: 17px;
  top: 11px;
  z-index: 10100;
}

.c0 .host-layout-header {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  height: 50px;
}

.c0 .host-layout-header h2 {
  position: relative;
  bottom: -1px;
  font-size: 1rem;
  line-height: 0.8rem;
}

.c0 .components {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100%;
}

.c0 .components .left {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  min-width: 200px;
  background-color: var(--primary-background-color);
  box-shadow: 0px 4px 12px rgba(0,0,0,0.25);
}

.c0 .components .right {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  position: relative;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 486px;
  padding: 40px;
  background-color: var(--primary-background-color);
  position: relative;
}

.c4 > div:not(:last-child) {
  margin-bottom: 24px;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  line-height: 29px;
}

.c6 .join-code-title {
  font-size: 18px;
  color: var(--primary-text-color);
  background-color: var(--primary-background-color);
  font-weight: 700;
  width: 97px;
  height: 21px;
}

.c6 .join-code {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  font-size: 56px;
  font-weight: bold;
  color: var(--secondary-text-color);
  background-color: var(--secondary-background-color);
  width: 305px;
  height: 70px;
}

.c5 {
  display: block;
  border: 2px solid var(--primary-text-color);
  font-size: 12px;
  font-weight: bold;
  padding: 0.7em;
  background: var(--secondary-background-color);
  color: var(--secondary-text-color);
}

.c5:hover,
.c5:active,
.c5:focus {
  background: var(--accent-background-color);
  cursor: pointer;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 100%;
}

.c3 .player-list-wrapper {
  overflow-x: auto;
  border-bottom: 1px solid var(--primary-text-color);
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c3 .buttons-wrapper {
  display: block;
  margin-bottom: 20px;
}

.c3 .bottom-left-wrapper {
  margin: 20px auto 32px;
  position: relative;
  bottom: 0;
  left: 0;
  right: 0;
}

.c3 .join-code-wrapper {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c3 .host-pregame-button {
  padding: 0;
  background-color: var(--primary-background-color);
  color: var(--primary-text-color);
  margin: auto auto 12px;
}

.c3 .start-carousing {
  font-size: 0.875rem;
  padding: 16px 34px;
}

.c3 .close-game {
  font-size: 0.625rem;
  padding: 8px 20px;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: absolute;
  top: 0;
  bottom: 0;
  right: 0;
  left: 0;
}

.c7 .game-description {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  background-color: var(--primary-background-color);
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  margin: 24px;
  padding: 8px;
  border-radius: 5px;
  font-size: 1.8rem;
  font-weight: 700;
  box-shadow: 0px 4px 12px rgba(0,0,0,0.25);
}

.c7 .game-description p {
  margin: 16px 0;
}

.c7 .top {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 60%;
}

.c7 .bottom {
  height: 40%;
}

@media (min-width:800px) {
  .c0 .host-layout-header {
    height: 130px;
  }

  .c0 .components .left {
    min-width: 450px;
  }

  .c0 .components .host-layout-header h2 {
    font-size: 2.5rem;
    margin-bottom: -4px;
    line-height: 2rem;
  }
}

@media (min-width:2000px) {
  .c0 .components .left {
    min-width: 700px;
  }

  .c0 .components .host-layout-header h2 {
    font-size: 3.5rem;
    margin-bottom: -35px;
    line-height: 7rem;
  }
}

@media (min-width:3500px) {
  .c0 .host-layout-header {
    height: 200px;
  }

  .c0 .components .left {
    min-width: 1100px;
  }

  .c0 .components .host-layout-header h2 {
    font-size: 6rem;
    margin-bottom: -25px;
    line-height: 7rem;
  }
}

<div
  className="c0 primary-background"
>
  <div
    className="hamburger-container"
  >
    <button
      className="c1"
      onClick={[Function]}
    />
  </div>
  <div
    className="components"
  >
    <div
      className="left"
    >
      <header
        className="c2 host-layout-header"
      >
        <h2>
          CARDS OF CAROUSAL
        </h2>
      </header>
      <div
        className="c3"
      >
        <div
          className="player-list-wrapper"
        >
          <div
            className="c4"
            data-testid="playerList-container"
          />
        </div>
        <div
          className="bottom-left-wrapper"
        >
          <div
            className="buttons-wrapper"
          >
            <button
              className="c5 host-pregame-button"
              onClick={[Function]}
            >
              <div
                className="start-carousing"
              >
                START CAROUSING
              </div>
            </button>
            <button
              className="c5 host-pregame-button"
              onClick={[Function]}
            >
              <div
                className="close-game"
              >
                CLOSE GAME
              </div>
            </button>
          </div>
          <div
            className="join-code-wrapper"
          >
            <div
              className="c6"
            >
              <p
                className="join-code-title"
              >
                JOIN CODE:
              </p>
              <p
                className="join-code"
                data-testid="join-code"
              >
                
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="right"
    >
      <div
        className="c7"
      >
        <div
          className="top"
        >
          <div
            className="game-description"
          >
            <p>
              Cards of Carousal is a game for lorem ipsum dolor.
            </p>
            <p>
              Lorem ipsum dolor sit amet consectetur adipisicing elit. Esse, doloribus quaerat incidunt excepturi odit eos!
            </p>
            <p>
              Qui delectus laboriosam aperiam maxime optio, architecto asperiores, at ullam.
            </p>
          </div>
        </div>
        <div
          className="bottom"
        >
          <div
            data-testid="game-settings"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
